base/set.jl

```diff
@@ -474,3 +474,220 @@ end
+
+# Examples
+```jldoctest
+julia> A = [1, 2, 3, 1];
+
+julia> replace!(isodd, A, 0, count=2)
+4-element Array{Int64,1}:
+ 0
+ 2
+ 0
+ 1
+```
+"""
+replace!(pred::Callable, A, new; count::Integer=typemax(Int)) =
+    _replace!(x -> if pred(x) Some(new) end, A, count)
+
+"""
+    replace!(prednew::Function, A; [count::Integer])
+
+For each value `x` in `A`, `prednew(x)` is called and must
+return either `nothing`, in which case no replacement occurs,
+or a value, possibly wrapped as a [`Some`](@ref) object, which
+will be used as a replacement for `x`.
+
+# Examples
+```jldoctest
+julia> replace!(Dict(1=>2, 3=>4)) do kv
+           if first(kv) < 3; first(kv)=>3 end
+       end
+Dict{Int64,Int64} with 2 entries:
+  3 => 4
+  1 => 3
+```
+
+!!! note
+    When `A` is an `AbstractDict` or `AbstractSet` collection, if
+    there are collisions among old and newly created keys, the result
+    depends on the iteration order (which can be undefined).
+    For example, the following may appear unexpected:
+
+```jldoctest
+julia> replace!(x->2x, Set([3, 6]))
+Set([12])
+```
+"""
+replace!(prednew::Callable, A; count::Integer=typemax(Int)) = _replace!(prednew, A, count)
+
+_replace(prednew::Callable, A, count::Integer) =
+    _replace(prednew, A, clamp(count, typemin(Int), typemax(Int)) % Int)
+
+
+"""
+    replace(A, old_new::Pair...; [count::Integer])
+
+Return a copy of collection `A` where, for each pair `old=>new` in `old_new`,
+all occurrences of `old` are replaced by `new`.
+If `count` is specified, then replace at most `count` occurrences in total.
+See also [`replace!`](@ref).
+
+# Examples
+```jldoctest
+julia> replace([1, 2, 1, 3], 1=>0, 2=>4, count=2)
+4-element Array{Int64,1}:
+ 0
+ 4
+ 1
+ 3
+```
+"""
+replace(A, old_new::Pair...; count::Integer=typemax(Int)) =
+    _replace!(copy(A), eltype(A), count, old_new)
+
+"""
+    replace(pred::Function, A, new; [count::Integer])
+
+Return a copy of collection `A` where all occurrences `x` for which
+`pred(x)` is true are replaced by `new`.
+
+# Examples
+```jldoctest
+julia> replace(isodd, [1, 2, 3, 1], 0, count=2)
+4-element Array{Int64,1}:
+ 0
+ 2
+ 0
+ 1
+```
+"""
+replace(pred::Callable, A, new; count::Integer=typemax(Int)) =
+    _replace!(x -> if pred(x) Some(new) end, copy(A), count)
+
+"""
+    replace(prednew::Function, A; [count::Integer])
+
+Return a copy of `A` where for each value `x` in `A`, `prednew(x)` is called
+and must return  either `nothing`, in which case no replacement occurs,
+or a value, possibly wrapped as a [`Some`](@ref) object, which
+will be used as a replacement for `x`.
+
+# Examples
+```jldoctest
+julia> replace(Dict(1=>2, 3=>4)) do kv
+           if first(kv) < 3; first(kv)=>3 end
+       end
+Dict{Int64,Int64} with 2 entries:
+  3 => 4
+  1 => 3
+```
+"""
+replace(prednew::Callable, A; count::Integer=typemax(Int)) = _replace!(prednew, copy(A), count)
+
+# Handle ambiguities
+replace!(a::Callable, b::Pair; count::Integer=-1) = throw(MethodError(replace!, a, b))
```

Comment:
Should be `(a, b)`. Else you get a `MethodError`, but it happens when failing to create the `MethodError`. ;-)

But should these really be errors? It's perfectly valid to have an array of pairs and to replace some pairs with others.

Reply:
Ah yes right! Those were not tested, just inserted to silence some ambiguities. You are right that what you said is perfectly valid, but this method doesn't have an array as parameter! This is just a silly signature, but Julia needs it!